### PR TITLE
[FLINK-21979] Cleanup completed checkpoint store after dispatcher cleans up HA services

### PR DESCRIPTION
**Pull request by dmvk**
## What is the purpose of the change

* Cleaning up HA services before CompletedCheckpointStore

## Brief change log

*(for example:)*
  - CompletedCheckpointStore is cleaned in scheduler's `closeAsync`, instead of on execution graph termination.

## Verifying this change

This change added tests and can be verified as follows:

- Integration test for a failover scenario described in JIRA.
- 
## Does this pull request potentially affect one of the following parts:

  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes)

### CODE REVIEW

**Review by tillrohrmann — CHANGES_REQUESTED** [review]
Hmm, there are quite a few test failures @dmvk. Could you double check whether they are caused by your changes?

**Comment by dmvk** [comment]
@tillrohrmann looking into it

**Comment by dmvk** [comment]
@tillrohrmann Fixed, we have a green build 🎉

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobRecoveryITCase.java
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobRecoveryITCase.java:61]: Is this a left-over from some testing?

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobRecoveryITCase.java
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobRecoveryITCase.java:77]: It looks as if we should revert this change.

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ExecutionGraphCheckpointCoordinatorTest.java
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ExecutionGraphCheckpointCoordinatorTest.java:71]: I think these tests are no longer correct. The test name says that the CC should be shut down on an `ExecutionGraph` failure. However, this is no longer true. It's the same for the other tests.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:786]: The else branch should be covered by `cleanUpJobGraph()`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:385]: Maybe we can introduce a helper method for this to give this operation a good name.

**Review by tillrohrmann — CHANGES_REQUESTED** [review]
Thanks for creating this PR @dmvk. I think the change looks already very good. 

I think we should take a look at a how we obtain the leading `JobMasterGateway` in the `DispatcherFailoverITCase`. At the moment, the solution is susceptible to memory visibility  issues.

What I would also suggest is to add a few more unit tests for the changed behaviour:

* A unit test that tests the shutdown behaviour of the `SchedulerBase` (e.g. make sure that the checkpoint services are shut down on `closeAsync`)
* A unit test making sure that the `SchedulerBase` is shut down if the `JobMaster` stops if such a test is not existing
* A unit test guarding the new `cleanUpJobGraph` and `cleanUpRemainingJobData` contract. One could instantiate a `TestingJobMasterService` that signals that its job is finished and then we should make sure that we first remove the job from the `JobGraphStore`, then close the `JobMasterService` and then clean up the remaining HA bookkeeping data.

Apart from that I had a couple of minor comments. PTAL.

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/JobMasterGatewayTester.java
Comment (dmvk) [reply flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/JobMasterGatewayTester.java:183]: 👍 We don't, it's not used anymore

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobRecoveryITCase.java
Comment (dmvk) [reply flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobRecoveryITCase.java:77]: 👍

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherFailoverITCase.java
Comment (dmvk) [reply flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherFailoverITCase.java:150]: fixed

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ExecutionGraphCheckpointCoordinatorTest.java
Comment (dmvk) [reply flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ExecutionGraphCheckpointCoordinatorTest.java:71]: I've renamed the test to `DefaultSchedulerCheckpointCoordinatorTest` and renamed tests so they reflect current behavior. Also I've added test for auto suspending of the running job.

**Comment by dmvk** [comment]
> A unit test that tests the shutdown behaviour of the SchedulerBase (e.g. make sure that the checkpoint services are shut down on closeAsync)

Stressed by `DefaultSchedulerCheckpointCoordinatorTest`

> A unit test making sure that the SchedulerBase is shut down if the JobMaster stops if such a test is not existing

This is stressed by `org.apache.flink.runtime.jobmaster.JobMasterTest#testJobMasterOnlyTerminatesAfterTheSchedulerHasClosed`

> A unit test guarding the new cleanUpJobGraph and cleanUpRemainingJobData contract. One could instantiate a TestingJobMasterService that signals that its job is finished and then we should make sure that we first remove the job from the JobGraphStore, then close the JobMasterService and then clean up the remaining HA bookkeeping data.

Stressed by:
```
org.apache.flink.runtime.dispatcher.DispatcherTest#testJobDataAreCleanedUpInCorrectOrderOnFinishedJob
org.apache.flink.runtime.dispatcher.DispatcherTest#testJobDataAreCleanedUpInCorrectOrderOnFailedJob
org.apache.flink.runtime.dispatcher.DispatcherTest#testFailingJobManagerRunnerCleanup
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java
Comment (dmvk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:384]: 👍

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/AbstractDispatcherTest.java
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/AbstractDispatcherTest.java:63]: I hope that this timeout is long enough for our CI infrastructure.

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherFailoverITCase.java
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherFailoverITCase.java:152]:
```suggestion
        jobGraphRemovalErrorReceived.await();
```

is a bit more robust on our CI infrastructure.

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherFailoverITCase.java
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherFailoverITCase.java:163]: I have to admit that I was a bit thrown off why this is working. Then I realized that `leaderElectionService.stop()` clears the `contender` field. One way to make it clearer that the old `Dispatcher` and its `JobMaster` are no longer used is to explicitly close it after the error is received.

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherTest.java
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherTest.java:164]: Instead of doing this, would it make sense to initialize the `haServices` field with a `StandaloneJobGraphStore`?

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherTest.java
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherTest.java:180]: I think this should go after the if block because the dispatcher might still use some of the services terminated in `super.tearDown()`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherTest.java
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherTest.java:645]: We could introduce constants for these names. Then it is easier to maintain them.

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherTest.java
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherTest.java:880]: Same here with the constants.

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/JobMasterTester.java
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/JobMasterTester.java:70]: I'm wondering whether this timeout is too low for our CI infrastructure.

Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/JobMasterTester.java:70]: Since this timeout is used for the rpc calls, I think it totally fine to set it to 1 minute w/o much harm.

**Review by tillrohrmann — CHANGES_REQUESTED** [review]
Thanks a lot for updating this PR @dmvk. The changes look really good. I had some last minor comments. I think that we can merge this PR after they are resolved.

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/AbstractDispatcherTest.java
Comment (dmvk) [reply flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/AbstractDispatcherTest.java:63]: increased to 1m

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherTest.java
Comment (dmvk) [reply flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherTest.java:164]: 👍

File: flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherFailoverITCase.java
Comment (dmvk) [reply flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherFailoverITCase.java:163]: I'm not able to fully close the first dispatcher here as the error would be thrown again during job graph removal. I guess this is also closer to a real world scenario, where other services wouldn't be closed at all.

I'll try to make the behavior of `leaderElectionService.stop()` bit more explicit with a comment.

Comment (dmvk) [reply flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherFailoverITCase.java:163]: I've also added call to `leaderElectionService.notLeader();`, so it's clear that we revoke leadership of the first dispatcher.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java
Comment (dmvk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:781]: 👍 I think that's a reasonable trade-off for now (until FLINK-11813 is resolved). If we were not able to remove job graph, we're up for the bigger troubles anyway :)

Dispatcher tests seem to pass, so let's see if we introduce any test failure in e2e tests by this.

**Review by tillrohrmann — APPROVED** [review]
Thanks for updating this PR @dmvk. LGTM. Merging this PR now.

**Pull request closed without merging**
